src/proofs/weighted_inner_product.rs

```diff
@@ -0,0 +1,915 @@
+#![allow(non_snake_case)]
+
+/*
+
+Copyright 2018 by Kzen Networks
+
+This file is part of bulletproof library
+(https://github.com/KZen-networks/bulletproof)
+
+bulletproof is free software: you can redistribute
+it and/or modify it under the terms of the GNU General Public
+License as published by the Free Software Foundation, either
+version 3 of the License, or (at your option) any later version.
+
+@license GPL-3.0+ <https://github.com/KZen-networks/bulletproof/blob/master/LICENSE>
+*/
+
+// based on the paper: https://eprint.iacr.org/2017/1066.pdf
+use curv::arithmetic::traits::Modulo;
+use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+use curv::cryptographic_primitives::hashing::traits::*;
+use curv::elliptic::curves::traits::*;
+use curv::BigInt;
+use curv::{FE, GE};
+
+use Errors::{self, WeightedInnerProdError};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WeightedInnerProdArg {
+    L: Vec<GE>,
+    R: Vec<GE>,
+    a_tag: GE,
+    b_tag: GE,
+    r_prime: BigInt,
+    s_prime: BigInt,
+    delta_prime: BigInt,
+}
+
+impl WeightedInnerProdArg {
+    pub fn prove(
+        G: &[GE],
+        H: &[GE],
+        g: &GE,
+        h: &GE,
+        P: &GE,
+        a: &[BigInt],
+        b: &[BigInt],
+        alpha: &BigInt,
+        mut L_vec: Vec<GE>,
+        mut R_vec: Vec<GE>,
+    ) -> WeightedInnerProdArg {
+        let n = G.len();
+        let order = FE::q();
+
+        // All of the input vectors must have the same length.
+        assert_eq!(G.len(), n);
+        assert_eq!(H.len(), n);
+        assert_eq!(a.len(), n);
+        assert_eq!(b.len(), n);
+        assert!(n.is_power_of_two());
+
+        // Generate scalar `y` uniformly from the scalar field
+        let y: BigInt = HSha256::create_hash_from_slice("Seed string decided by P,V!".as_bytes());
+        let powers_y = scalar_powers(y.clone(), n);
```

Comment:
optimisation: I saw how `scalar_powers` are implemented, basically storing a vector if y^i. This takes a lot of space, we should consider how it compares to compute the scalar powers on the fly each time we use them